File: languages/go/client/integrity.go
package client

import (
	"context"
	"errors"

	"github.com/bloock/bloock-sdk-go/v2/entity/integrity"
	"github.com/bloock/bloock-sdk-go/v2/entity/record"
	"github.com/bloock/bloock-sdk-go/v2/internal/bridge"
	"github.com/bloock/bloock-sdk-go/v2/internal/bridge/proto"
	"github.com/bloock/bloock-sdk-go/v2/internal/config"
)

// IntegrityClient provides functionality to interact with the [Bloock Integrity service].
//
// [Bloock Integrity service]: https://bloock.com
type IntegrityClient struct {
	bridgeClient bridge.BloockBridge
	configData   *proto.ConfigData
}

// NewIntegrityClient creates a new IntegrityClient with default configuration.
func NewIntegrityClient() IntegrityClient {
	return IntegrityClient{
		bridgeClient: bridge.NewBloockBridge(),
		configData:   config.NewConfigDataDefault(),
	}
}

// NewIntegrityClientWithConfig creates a new IntegrityClient with the given configuration.
func NewIntegrityClientWithConfig(configData *proto.ConfigData) IntegrityClient {
	return IntegrityClient{
		bridgeClient: bridge.NewBloockBridge(),
		configData:   configData,
	}
}

// SendRecords sends records to the Bloock Integrity service for certification.
func (c *IntegrityClient) SendRecords(records []record.Record) ([]integrity.RecordReceipt, error) {
	res, err := c.bridgeClient.Integrity().SendRecords(context.Background(), &proto.SendRecordsRequest{
		ConfigData: c.configData,
		Records:    record.MapRecordsToProto(records),
	})

	if err != nil {
		return []integrity.RecordReceipt{}, err
	}

	if res.Error != nil {
		return []integrity.RecordReceipt{}, errors.New(res.Error.Message)
	}

	receipts := make([]integrity.RecordReceipt, len(res.Records))
	for i, record := range res.Records {
		receipts[i] = integrity.NewRecordReceiptFromProto(record)
	}

	return receipts, nil
}

// GetAnchor retrieves an anchor by its ID from the Bloock Integrity service.
func (c *IntegrityClient) GetAnchor(anchorID int64) (integrity.Anchor, error) {
	res, err := c.bridgeClient.Integrity().GetAnchor(context.Background(), &proto.GetAnchorRequest{
		ConfigData: c.configData,
		AnchorId:   anchorID,
	})

	if err != nil {
		return integrity.Anchor{}, err
	}

	if res.Error != nil {
		return integrity.Anchor{}, errors.New(res.Error.Message)
	}

	return integrity.NewAnchorFromProto(res.Anchor), nil
}

// WaitAnchor waits for the completion of an anchor on the Bloock Integrity service.
func (c *IntegrityClient) WaitAnchor(anchorID int64, params integrity.AnchorParams) (integrity.Anchor, error) {
	if params.Timeout == 0 {
		params.Timeout = int64(120000)
	}

	res, err := c.bridgeClient.Integrity().WaitAnchor(context.Background(), &proto.WaitAnchorRequest{
		ConfigData: c.configData,
		AnchorId:   anchorID,
		Timeout:    params.Timeout,
	})

	if err != nil {
		return integrity.Anchor{}, err
	}

	if res.Error != nil {
		return integrity.Anchor{}, errors.New(res.Error.Message)
	}

	return integrity.NewAnchorFromProto(res.Anchor), nil
}

// GetProof retrieves a proof for a set of records from the Bloock Integrity service.
func (c *IntegrityClient) GetProof(records []record.Record) (integrity.Proof, error) {
	res, err := c.bridgeClient.Integrity().GetProof(context.Background(), &proto.GetProofRequest{
		ConfigData: c.configData,
		Records:    record.MapRecordsToProto(records),
	})

	if err != nil {
		return integrity.Proof{}, err
	}

	if res.Error != nil {
		return integrity.Proof{}, errors.New(res.Error.Message)
	}

	return integrity.NewProofFromProto(res.Proof), nil
}

// VerifyProof verifies the integrity of a proof.
func (c *IntegrityClient) VerifyProof(proof integrity.Proof) (string, error) {
	res, err := c.bridgeClient.Integrity().VerifyProof(context.Background(), &proto.VerifyProofRequest{
		ConfigData: c.configData,
		Proof:      proof.ToProto(),
	})

	if err != nil {
		return "", err
	}

	if res.Error != nil {
		return "", errors.New(res.Error.Message)
	}

	return *res.Record, nil
}

// VerifyRecords verifies the integrity of a set of records.
func (c *IntegrityClient) VerifyRecords(records []record.Record, params integrity.NetworkParams) (uint64, error) {
	res, err := c.bridgeClient.Integrity().VerifyRecords(context.Background(), &proto.VerifyRecordsRequest{
		ConfigData: c.configData,
		Records:    record.MapRecordsToProto(records),
		Network:    integrity.NetworkToProto(params.Network),
	})

	if err != nil {
		return 0, err
	}

	if res.Error != nil {
		return 0, errors.New(res.Error.Message)
	}

	return res.Timestamp, nil
}

// ValidateRoot validates the integrity of a merkle root proof on blockchain.
func (c *IntegrityClient) ValidateRoot(root string, params integrity.NetworkParams) (uint64, error) {
	res, err := c.bridgeClient.Integrity().ValidateRoot(context.Background(), &proto.ValidateRootRequest{
		ConfigData: c.configData,
		Root:       root,
		Network:    *integrity.NetworkToProto(params.Network),
	})

	if err != nil {
		return 0, err
	}

	if res.Error != nil {
		return 0, errors.New(res.Error.Message)
	}

	return res.Timestamp, nil
}

File: languages/js/src/client/integrity.ts
import { BloockBridge } from "../bridge/bridge";
import { ConfigData } from "../bridge/proto/bloock_config";
import {
  GetAnchorRequest,
  GetProofRequest,
  SendRecordsRequest,
  ValidateRootRequest,
  VerifyProofRequest,
  VerifyRecordsRequest,
  WaitAnchorRequest
} from "../bridge/proto/bloock_integrity";
import { NewConfigData } from "../config/config";
import { Network, Proof, RecordReceipt } from "../entity/integrity";
import { Anchor } from "../entity/integrity/anchor";
import { Record } from "../entity/record";

/**
 * Provides functionality to interact with the [Bloock Integrity service](https://dashboard.bloock.com/login).
 */
export class IntegrityClient {
  private bridge: BloockBridge;
  private configData: ConfigData | undefined;

  /**
   * Creates a new IntegrityClient with default configuration.
   * @param configData
   */
  constructor(configData?: ConfigData) {
    this.bridge = new BloockBridge();
    this.configData = NewConfigData(configData);
  }

  /**
   * Sends records to the Bloock Integrity service for certification.
   * @param records
   * @returns
   */
  public async sendRecords(records: Record[]): Promise<RecordReceipt[]> {
    const request = SendRecordsRequest.fromPartial({
      configData: this.configData,
      records: records.map(x => x.toProto())
    });

    return this.bridge
      .getIntegrity()
      .SendRecords(request)
      .then(res => {
        if (res.error) {
          throw res.error;
        }
        return res.records.map(r => RecordReceipt.fromProto(r));
      });
  }

  /**
   * Gets an anchor by its ID from the Bloock Integrity service.
   * @param anchorId
   * @returns
   */
  public async getAnchor(anchorId: number): Promise<Anchor> {
    const request = GetAnchorRequest.fromPartial({
      configData: this.configData,
      anchorId: anchorId
    });

    return this.bridge
      .getIntegrity()
      .GetAnchor(request)
      .then(res => {
        if (res.error) {
          throw res.error;
        }
        return Anchor.fromProto(res.anchor!);
      });
  }

  /**
   * Waits for the completion of an anchor on the Bloock Integrity service.
   * @param anchorId
   * @param timeout
   * @returns
   */
  public async waitAnchor(anchorId: number, timeout?: number): Promise<Anchor> {
    const request = WaitAnchorRequest.fromPartial({
      configData: this.configData,
      anchorId: anchorId,
      timeout: timeout !== null ? timeout : 120000
    });

    return this.bridge
      .getIntegrity()
      .WaitAnchor(request)
      .then(res => {
        if (res.error) {
          throw res.error;
        }
        return Anchor.fromProto(res.anchor!);
      });
  }

  /**
   * Gets a proof for a set of records from the Bloock Integrity service.
   * @param records
   * @returns
   */
  public async getProof(records: Record[]): Promise<Proof> {
    const request = GetProofRequest.fromPartial({
      configData: this.configData,
      records: records.map(x => x.toProto())
    });

    return this.bridge
      .getIntegrity()
      .GetProof(request)
      .then(res => {
        if (res.error) {
          throw res.error;
        }
        return Proof.fromProto(res.proof!);
      });
  }

  /**
   * Validates the integrity of a merkle root proof on blockchain.
   * @param root
   * @param network
   * @returns
   */
  public async validateRoot(root: string, network: Network): Promise<number> {
    const request = ValidateRootRequest.fromPartial({
      configData: this.configData,
      root: root,
      network: network
    });

    return this.bridge
      .getIntegrity()
      .ValidateRoot(request)
      .then(res => {
        if (res.error) {
          throw res.error;
        }
        return res.timestamp;
      });
  }

  /**
   * Verifies the integrity of a proof.
   * @param proof
   * @returns
   */
  public async verifyProof(proof: Proof): Promise<string> {
    const request = VerifyProofRequest.fromPartial({
      configData: this.configData,
      proof: proof.toProto()
    });

    return this.bridge
      .getIntegrity()
      .VerifyProof(request)
      .then(res => {
        if (res.error) {
          throw res.error;
        }
        return res.record!;
      });
  }

  /**
   * Verifies the integrity of a set of records.
   * @param records
   * @param network
   * @returns
   */
  public async verifyRecords(
    records: Record[],
    network?: Network
  ): Promise<number> {
    const request = VerifyRecordsRequest.fromPartial({
      configData: this.configData,
      records: records.map(x => x.toProto()),
      network: network
    });

    return this.bridge
      .getIntegrity()
      .VerifyRecords(request)
      .then(res => {
        if (res.error) {
          throw res.error;
        }
        return res.timestamp;
      });
  }
}
